### PR TITLE
Add index select backward

### PR DESCRIPTION
**Pull request by AdvancedCompiler**
### PR Category
Operator

### Type of Change
 New Feature 

### Description
Implement index_select_backward operator

### Issue
#316 

### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [ ] Change is fully covered by a UT.

### Performance
index_select_backward_pref
### Accuracy

index_select_backward_accuracy

### CODE REVIEW

File: src/flag_gems/ops/index_select.py [hunk 1]
```diff
@@ -62,3 +62,59 @@ def index_select(inp, dim, index):
...
+    assert index.ndim <= 1, "Index should have dimension 1 or 0"
+    for i in index:
+        assert i >= 0 and i < self_sizes[dim], "Index out of range"
+    if index.ndim == 0:
+        index = index.unsqueeze(0)
+    index_shape = list(index.shape)
+    dim = dim % len(self_sizes)
+    grad_shape = list(grad.shape)
+    assert grad_shape[dim] == index_shape[0], "Index out of range"
+    grad = dim_compress(grad, dim)
```
Comment (iclementine) [inline src/flag_gems/ops/index_select.py]: Function `dim_compress` move the specified dim to the inner-most dimension and make a contiguous tensor, which is designed to be used in reduction/scan operator when other dimensions are considered batch dimensions and the reduction operation is performed on each 1d sub tensors, which will be loaded and iterated over. Thus being contiguous is good for this case.

For the backward of `index_select`, it is actually the opposite. Dimension `dim` is used  as an indexing dimension. Then several `n-1`d subtensors are inserted into a zeros tensor. So, being contiguous on dimension `dim` is not what we want. We rather want it to be the outer-most dimension.

Comment (shuailong616) [reply src/flag_gems/ops/index_select.py]: Ok,we will change it.

**Review by Bowen12992 — DISMISSED** [review]
LGTM for CI coverage

File: src/flag_gems/ops/index_select.py [hunk 2]
```diff
@@ -62,3 +63,74 @@ def index_select(inp, dim, index):
...
+
+def dim_compress_backward(inp, dims):
+    if isinstance(dims, int):
+        dims = [dims]
+    dim = inp.ndim
+    stride = inp.stride()
+    batch_dim = [i for i in range(dim) if i not in dims]
+    sorted_reduction_dim = sorted(dims, key=lambda x: stride[x], reverse=True)
+    order = sorted_reduction_dim + batch_dim
+    return inp.permute(order).contiguous()
```
Comment (iclementine) [inline src/flag_gems/ops/index_select.py:83]: looks good.

File: src/flag_gems/ops/index_select.py [hunk 3]
```diff
@@ -62,3 +63,74 @@ def index_select(inp, dim, index):
...
+        tl.store(out + out_off, selected, mask=grad_mask)
+
+
+# function
+def index_select_backward(grad, self_sizes, dim, index):
+    logging.debug("GEMS INDEX SELECT BACKWARD")
+    assert dim >= -len(self_sizes) and dim < len(self_sizes), "Invalid dim"
+    assert index.ndim <= 1, "Index should have dimension 1 or 0"
+    for i in index:
+        assert i >= 0 and i < self_sizes[dim], "Index out of range"
```
Comment (iclementine) [inline src/flag_gems/ops/index_select.py]: I suggest removing this out-of-bound checking, since it involves slicing & compare for each index, which means a large overhead.

File: tests/test_reduction_ops.py [hunk 4]
```diff
@@ -689,6 +689,36 @@ def test_accuracy_index_select(shape, dim, dtype):
...
+@pytest.mark.parametrize("dim", DIM_LIST)
+@pytest.mark.parametrize("dtype", FLOAT_DTYPES)
+def test_accuracy_index_select_backward(shape, dim, dtype):
+    inp = torch.randn(shape, dtype=dtype, device="cuda", requires_grad=True)
+    ref_inp = to_reference(inp)
+    from math import floor
+
+    index_size = inp.size(dim)
+    index = torch.randint(0, index_size, [floor(index_size * 0.8)], device="cuda")
+    index = torch.unique(index)
```
Comment (iclementine) [inline tests/test_reduction_ops.py]: What if there is duplicated index? I don't think current implementation can handle this.

File: src/flag_gems/ops/index_select.py [hunk 5]
```diff
@@ -75,3 +77,74 @@ def index_select(inp, dim, index):
...
+        index = index.unsqueeze(0)
+    index_shape = list(index.shape)
+    dim = dim % len(self_sizes)
+    grad_shape = list(grad.shape)
+    assert grad_shape[dim] == index_shape[0], "Index out of range"
+    grad = dim_compress_backward(grad, dim)
+    grad_shape = list(grad.shape)
+    out_shape = list(grad.shape)
+    tem_shape = list(grad.shape[1:])
+    tem_shape[-1] = 1
```
Comment (iclementine) [inline src/flag_gems/ops/index_select.py]: What does this shape `tmp_shape` intended to be ?

Comment (iclementine) [reply src/flag_gems/ops/index_select.py]: Maybe choose a more semantically meaningful name.

File: src/flag_gems/ops/index_select.py (hunk 3 above)
Comment (shuailong616) [reply src/flag_gems/ops/index_select.py]: ok

File: tests/test_reduction_ops.py (hunk 4 above)
Comment (shuailong616) [reply tests/test_reduction_ops.py]: Yes, for duplicted index , torch.autograd.grad can handle this case, our design code also can handle. But for accuracy test,  out_grad and index random generate ,not meet same index and value keep correspond's condition. So, we add the unique statement.；

 index


out_grad


normally ,the index is duplicated,the correwponding value overwritten, and the result is correct.

Comment (iclementine) [reply tests/test_reduction_ops.py]: No, I don't see how it can handle duplicated indices. With duplicated indices, a sub-tensor is selected multiple times in the forward op, thus the corresponding gradient should accumulate into the corresponding input gradient.

There is no such accumulation in the kernel. I add a test, and it always fails.

```python
@pytest.mark.index_select_backward
@pytest.mark.parametrize("shape", REDUCTION_SHAPES)
@pytest.mark.parametrize("dim", DIM_LIST)
@pytest.mark.parametrize("dtype", FLOAT_DTYPES)
def test_accuracy_index_select_backward(shape, dim, dtype):
    inp = torch.randn(shape, dtype=dtype, device="cuda", requires_grad=True)
    ref_inp = to_reference(inp)
    from math import floor

    index_size = inp.size(dim)
    index = torch.tensor([0, 0, 0, 0], device="cuda")
    # index = torch.unique(index)
    if len(index) == 0:
        pass
    else:
        ref_index = to_reference(index)
        ref_out = torch.index_select(ref_inp, dim, ref_index)
        out_grad = torch.randn_like(ref_out)
        ref_grad = to_reference(out_grad)
        (ref_in_grad,) = torch.autograd.grad(ref_out, ref_inp, ref_grad)

        with flag_gems.use_gems():
            res_out = torch.index_select(inp, dim, index)
            (res_in_grad,) = torch.autograd.grad(res_out, inp, out_grad)
        res_out = to_reference(res_out)
        res_in_grad = to_reference(res_in_grad)
        gems_assert_equal(res_out, ref_out)
        gems_assert_equal(res_in_grad, ref_in_grad)
```

Comment (shuailong616) [reply tests/test_reduction_ops.py:703]: out_grad !=ref_out

Comment (AdvancedCompiler) [reply tests/test_reduction_ops.py]: ok

File: src/flag_gems/ops/index_select.py (hunk 5 above)
Comment (AdvancedCompiler) [reply src/flag_gems/ops/index_select.py]: ok

File: src/flag_gems/ops/index_select.py (hunk 1 above)
Comment (shuailong616) [reply src/flag_gems/ops/index_select.py]: Hello, there is a precision issue when calling atomic_add, and I'm not sure how to resolve it.

File: src/flag_gems/ops/index_select.py [hunk 6]
```diff
@@ -75,3 +77,82 @@ def index_select(inp, dim, index):
...
+    cols_offsets = pid_y * BLOCK_N + tl.arange(0, BLOCK_N)
+
+    grad_mask = (rows_offsets < M) and (cols_offsets < N)
+    indices = tl.load(index + rows_offsets, mask=(rows_offsets < M), other=0)
+
+    for i in range(0, num_blocks_per_CTA):
+        grad_off = (pid_x * num_blocks_per_CTA + i) * N + cols_offsets
+        out_off = (indices * num_blocks_per_CTA + i) * N + cols_offsets
+        selected = tl.load(grad + grad_off, mask=grad_mask, other=0.0)
+        tl.atomic_add(out + out_off, selected, mask=grad_mask)
```
Comment (iclementine) [inline src/flag_gems/ops/index_select.py:116]: Since this kernel uses `atomic_add` and also `autotune`, you should add `out` to `reset_to_zero` to avoid it being added to many times.

File: benchmark/test_select_and_slice_perf.py [hunk 7]
```diff
@@ -223,6 +223,30 @@ def select_scatter_input_fn(shape, dtype, device):
...
+        dim = 0
+        index_size = inp.size(dim)
+        from math import floor
+
+        index = torch.randint(
+            0, index_size, [floor(index_size * threshold)], device=device
+        )
+        yield inp, dim, index
+
+    bench = TensorSelectBenchmark(
```
Comment (iclementine) [inline benchmark/test_select_and_slice_perf.py]: You need to override

```python
    def get_gbps(self, args, latency=None):
        # """Return the dynamic input iterator for each Operator."""
        raise NotImplementedError(
            "Each Benchmark must implement its own input iterator."
        )
```

for it to compute gbps metric.

Comment (AdvancedCompiler) [reply benchmark/test_select_and_slice_perf.py]: ok

**Review by tengqm — APPROVED** [review]

**Review by w1120029931-bit — CHANGES_REQUESTED** [review]
please resolve conflicts
The code architecture has been adjusted,
test_reduction_ops.py
does not exist currently, Each operator requires separate functional test files and performance test files. Please refer to the relevant code modifications in the repository
AND
benchmark/test*_perf.py should be benchmark/test*.py

**Comment by w1120029931-bit** [comment]
Is the PR still being followed up
